<a href="https://colab.research.google.com/github/Lanlanb/Studycase/blob/main/Data_Wrangling_of_Hotel_bookings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Hotel Bookings

![Hotels Gif by Giphy](https://media4.giphy.com/media/fnQ9jP3cXYIj2AKjcV/giphy.gif?cid=6c09b952stu6jdxk2azuu224n0vws7pypbskjs2xwywk40fi&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=g)

# **⚠️ Disclaimer**
>
> Notebook ini merupakan bagian dari studi kasus portofolio yang saya kerjakan bersama dua teman saya, [Imelda](https://github.com/Imeldyyn) dan Rizki, di mana Imelda menyediakan dataset [`hotel_booking.csv`](https://raw.githubusercontent.com/Lanlanb/Studycase/refs/heads/main/hotel_bookings.csv) yang akan kami gunakan dalam EDA masing-masing, namun setiap dari kami bekerja secara independen dan menggunakan versi proyek yang berbeda.
>
> Proyek ini murni untuk tujuan pembelajaran, tanpa keterkaitan profesional dengan hotel atau layanan manapun. Saya pribadi melihat tantangan seperti kurangnya kelengkapan data dan proyek sebagai latihan yang bermanfaat untuk persiapan menghadapi skenario nyata di dunia kerja.

# **Peran Saya**
Dalam proyek ini, saya berperan sebagai seorang Data Analyst yang menerima tugas dari sebuah perusahaan untuk melakukan analisis data terkait reservasi hotel. Skenario ini mensimulasikan tantangan nyata di dunia kerja, di mana perusahaan, yang dalam hal ini bertindak sebagai klien, hanya memberikan dataset hotel_booking.csv tanpa penjelasan mendetail mengenai konteks atau variabel yang ada.

# 1) gathering data

In [ ]:
import pandas as pd

# Set agar semua kolom ditampilkan
pd.options.display.max_columns = None

In [ ]:
# read dataset
path = "https://raw.githubusercontent.com/Lanlanb/Studycase/refs/heads/main/hotel_bookings.csv"

data = pd.read_csv(path)

In [ ]:
data.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type  agent  company  \
0                  C                3   No Deposit    NaN      NaN   
1                  C                4   No Deposit    NaN      NaN   
2                  C                0   No Deposit    NaN      NaN   
3                  A                0   No Deposit  304.0      NaN   
4                  A                0   No Deposit  240.0      NaN   

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests reservation_status reservation_status_date  
0                          0          Check-Out                7/1/2015  
1                          0          Check-Out                7/1/2015  
2                          0          Check-Out                7/2/2015  
3                          0          Check-Out                7/2/2015  
4                          1          Check-Out                7/3/2015

In [ ]:
data.tail()

hotel  is_canceled  lead_time  arrival_date_year  \
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
119385                     5       2       0.0       0   BB     BEL   
119386                     5       3       0.0       0   BB     FRA   
119387                     5       2       0.0       0   BB     DEU   
119388                     5       2       0.0       0   BB     GBR   
119389                     7       2       0.0       0   HB     DEU   

       market_segment distribution_channel  is_repeated_guest  \
119385  Offline TA/TO                TA/TO                  0   
119386      Online TA                TA/TO                  0   
119387      Online TA                TA/TO                  0   
119388      Online TA                TA/TO                  0   
119389      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
119385                       0                               0   
119386                       0                               0   
119387                       0                               0   
119388                       0                               0   
119389                       0                               0   

       reserved_room_type assigned_room_type  booking_changes deposit_type  \
119385                  A                  A                0   No Deposit   
119386                  E                  E                0   No Deposit   
119387                  D                  D                0   No Deposit   
119388                  A                  A                0   No Deposit   
119389                  A                  A                0   No Deposit   

        agent  company  days_in_waiting_list customer_type     adr  \
119385  394.0      NaN                     0     Transient   96.14   
119386    9.0      NaN                     0     Transient  225.43   
119387    9.0      NaN                     0     Transient  157.71   
119388   89.0      NaN                     0     Transient  104.40   
119389    9.0      NaN                     0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
119385                            0                          0   
119386                            0                          2   
119387                            0                          4   
119388                            0                          0   
119389                            0                          2   

       reservation_status reservation_status_date  
119385          Check-Out                9/6/2017  
119386          Check-Out                9/7/2017  
119387          Check-Out                9/7/2017  
119388          Check-Out                9/7/2017  
119389          Check-Out                9/7/2017

# 2) assesing data

In [ ]:
data.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

Jika dilihat sekilas, variabel seperti children, country, agent dan company memiliki jumlah 'nilai yang tidak null' yang berbeda dibandingkan dengan variabel lainnya. Dengan selisih yang beragam di mana variabel company memiliki perbandingan 'nilai yang tidak null' terbanyak, maka hal tersebut dapat menjadi indikasi adanya ketidakseimbangan data yang disebabkan oleh adanya missing value.

In [ ]:
data.value_counts()

hotel         is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type    adr     required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel    0            256        2016               October             43                        16                         2                        3                     2       0.0       0       BB    DEU      Online TA       TA/TO                 0                  0                       0                               A                   A                   0                No Deposit    9.0    333.0    0                     Transient-Party  100.75  0                            0                          Check-Out           10/21/2016                 7
Resort Hotel  0            24         2015               November            45                        3                          3                        10                    1       0.0       0       BB    FRA      Corporate       Corporate             0                  0                       0                               A                   A                   2                No Deposit    334.0  281.0    0                     Transient-Party  40.00   0                            0                          Check-Out           11/16/2015                 5
                           36         2015               November            45                        7                          2                        6                     1       0.0       0       BB    DEU      Corporate       Corporate             0                  0                       0                               A                   A                   1                No Deposit    185.0  281.0    0                     Transient-Party  36.00   0                            0                          Check-Out           11/15/2015                 4
                                                                                                                                                                                                                 AUT      Corporate       Corporate             0                  0                       0                               A                   A                   1                No Deposit    185.0  281.0    0                     Transient-Party  36.00   0                            0                          Check-Out           11/15/2015                 3
City Hotel    0            0          2015               August              33                        13                         0                        2                     2       0.0       0       BB    PRT      Online TA       TA/TO                 0                  0                       0                               A                   B                   0                No Deposit    9.0    9.0      0                     Transient        85.00   0                            0                          Check-Out           8/15/2015                  2
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       ..
Resort Hotel  0

In [ ]:
data.nunique()

hotel                                2
is_canceled                          2
lead_time                          479
arrival_date_year                    3
arrival_date_month                  12
arrival_date_week_number            53
arrival_date_day_of_month           31
stays_in_weekend_nights             17
stays_in_week_nights                35
adults                              14
children                             5
babies                               5
meal                                 5
country                            177
market_segment                       8
distribution_channel                 5
is_repeated_guest                    2
previous_cancellations              15
previous_bookings_not_canceled      73
reserved_room_type                  10
assigned_room_type                  12
booking_changes                     21
deposit_type                         3
agent                              333
company                            352
days_in_waiting_list               128
customer_type                        4
adr                               8879
required_car_parking_spaces          5
total_of_special_requests            6
reservation_status                   3
reservation_status_date            926
dtype: int64

Dari penggunaan perintah di atas, jumlah nilai unik pada tiap variabel dalam dataset jadi diketahui. Selain itu, melalui informasi umum (`info`) setiap variabel, karakteristik masing-masing variabel dapat dilihat secara lebih rinci.

Sebagai contoh, mari lihat variabel terakhir dalam dataset:

In [ ]:
data['reservation_status_date'].describe()

count         119390
unique           926
top       10/21/2015
freq            1461
Name: reservation_status_date, dtype: object

Diambil juga lima sampel data secara acak dari variabel reservation_status_date untuk memberikan gambaran umum mengenai distribusi nilai dalam variabel tersebut:

In [ ]:
data['reservation_status_date'].sample(5)

60614     10/21/2015
95485      8/20/2016
111946     5/18/2017
68382      2/12/2017
66922     11/21/2016
Name: reservation_status_date, dtype: object

In [ ]:
# Mengatur jumlah baris yang ditampilkan di output
pd.set_option('display.max_rows', None)

In [ ]:
data['reservation_status_date']

Variabel `reservation_status_date` memiliki 926 nilai unik yang berisi informasi tanggal reservasi hotel. Jumlah kategori yang begitu banyak tampak kurang lazim, mengingat tipe data variabel ini adalah objek. Dari perspektif kualitas, ini berpotensi menurunkan efisiensi dalam proses pengolahan data dan analisis, terutama jika dataset tidak didampingi deskripsi yang jelas.

Untuk menghadapi situasi serupa, sangat penting memahami karakteristik setiap variabel dalam dataset sejak awal. Langkah sederhana seperti pertanyaan, “Apa sebenarnya makna dari variabel ini?” membantu mengarahkan proses analisis ke depan. Namun, Dalam proyek ini saya mencoba memposisikan diri sebagaimana peran yang sudah saya tetapkan sejak awal, yakni;


> "bagaimana-jika penyedia dataset dan/atau klien tidak menyediakan kelengkapan proyek atau dataset secara lengkap?"

Ini adalah skenario yang sangat mungkin terjadi di dunia nyata dan saya mengerjakan proyek ini sambil tetap mempertahankan peran dan pemikiran saya secara pribadi.

Untuk mengatasi masalah ini, saya memutuskan untuk membagi dataset menjadi dua bagian berbeda—yakni, berdasarkan tipe data variabelnya. Pemisahan ini bukanlah langkah final; melainkan hanya pendekatan awal yang bertujuan untuk lebih memahami struktur dataset secara lebih mendalam. Dengan memisahkan variabel berdasarkan tipenya, saya dapat melihat lebih jelas peran masing-masing variabel dalam konteks keseluruhan dataset. Pendekatan ini membantu saya untuk lebih terarah dalam melakukan analisis, sehingga proses pemrosesan data bisa berjalan dengan lebih lancar.

## Understanding data through variable division

### categorical

In [ ]:
# memisahkan dataset berdasarkan tipedatanya
categorical = data.select_dtypes(include=['object'])
numerical = data.select_dtypes(include=['int64', 'float64'])

categorical

hotel arrival_date_month meal country market_segment  \
0       Resort Hotel               July   BB     PRT         Direct   
1       Resort Hotel               July   BB     PRT         Direct   
2       Resort Hotel               July   BB     GBR         Direct   
3       Resort Hotel               July   BB     GBR      Corporate   
4       Resort Hotel               July   BB     GBR      Online TA   
...              ...                ...  ...     ...            ...   
119385    City Hotel             August   BB     BEL  Offline TA/TO   
119386    City Hotel             August   BB     FRA      Online TA   
119387    City Hotel             August   BB     DEU      Online TA   
119388    City Hotel             August   BB     GBR      Online TA   
119389    City Hotel             August   HB     DEU      Online TA   

       distribution_channel reserved_room_type assigned_room_type  \
0                    Direct                  C                  C   
1                    Direct                  C                  C   
2                    Direct                  A                  C   
3                 Corporate                  A                  A   
4                     TA/TO                  A                  A   
...                     ...                ...                ...   
119385                TA/TO                  A                  A   
119386                TA/TO                  E                  E   
119387                TA/TO                  D                  D   
119388                TA/TO                  A                  A   
119389                TA/TO                  A                  A   

       deposit_type customer_type reservation_status reservation_status_date  
0        No Deposit     Transient          Check-Out                7/1/2015  
1        No Deposit     Transient          Check-Out                7/1/2015  
2        No Deposit     Transient          Check-Out                7/2/2015  
3        No Deposit     Transient          Check-Out                7/2/2015  
4        No Deposit     Transient          Check-Out                7/3/2015  
...             ...           ...                ...                     ...  
119385   No Deposit     Transient          Check-Out                9/6/2017  
119386   No Deposit     Transient          Check-Out                9/7/2017  
119387   No Deposit     Transient          Check-Out                9/7/2017  
119388   No Deposit     Transient          Check-Out                9/7/2017  
119389   No Deposit     Transient          Check-Out                9/7/2017  

[119390 rows x 12 columns]

In [ ]:
categorical.value_counts()

hotel         arrival_date_month  meal       country  market_segment  distribution_channel  reserved_room_type  assigned_room_type  deposit_type  customer_type    reservation_status  reservation_status_date
City Hotel    May                 BB         PRT      Groups          TA/TO                 A                   A                   Non Refund    Transient        Canceled            11/25/2016                 225
              August              BB         PRT      Groups          TA/TO                 A                   A                   Non Refund    Transient-Party  Canceled            7/2/2015                   217
                                                                                                                                                  Contract         Canceled            1/1/2015                   180
              November            BB         PRT      Groups          TA/TO                 A                   A                   Non Refund    Transient        Canceled            4/4/2016                   180
                                                      Offline TA/TO   TA/TO                 A                   A                   Non Refund    Transient        Canceled            2/1/2016                   170
                                                                                                                                                                                                                 ... 
              March               BB         FRA      Offline TA/TO   TA/TO                 A                   D                   No Deposit    Transient        Check-Out           3/14/2017                    1
                                                                                                                                                                                       3/21/2016                    1
                                                                                                                                                                                       3/21/2017                    1
                                                                                                                                                                                       3/23/2017                    1
Resort Hotel  September           Undefined  PRT      Offline TA/TO   TA/TO                 D                   D                   No Deposit    Transient        Canceled            8/26/2016                    1
Name: count, Length: 68609, dtype: int64

In [ ]:
for col in categorical.columns:
    print("===============================")
    print(categorical[col].value_counts())
    print("\n")

hotel
City Hotel      79330
Resort Hotel    40060
Name: count, dtype: int64


arrival_date_month
August       13877
July         12661
May          11791
October      11160
April        11089
June         10939
September    10508
March         9794
February      8068
November      6794
December      6780
January       5929
Name: count, dtype: int64


meal
BB           92310
HB           14463
SC           10650
Undefined     1169
FB             798
Name: count, dtype: int64


country
PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
       ...  
DJI        1
BWA        1
HND        1
VGB        1
NAM        1
Name: count, Length: 177, dtype: int64


market_segment
Online TA        56477
Offline TA/TO    24219
Groups           19811
Direct           12606
Corporate         5295
Complementary      743
Aviation           237
Undefined            2
Name: count, dtype: int64


distribution_channel
TA/TO        97870
Direct       14645
Corporate     6677
GDS            193
Und

Dari hasil yang dikeluarkan, adapun definisi mengenai variabel yang dapat ditemukan:

```
hotel: Menunjukkan jenis hotel tempat pemesanan dilakukan. Ada dua jenis hotel yang terdaftar, yaitu "City Hotel" dan "Resort Hotel".
arrival_date_month: Menunjukkan bulan kedatangan tamu yang melakukan pemesanan.
meal: Menunjukkan jenis makanan yang dipesan oleh tamu. Misalnya, "BB" (Bed and Breakfast).
country: Negara asal tamu yang melakukan pemesanan.
market_segment: Segmentasi pasar yang menunjukkan kategori asal pemesanan, seperti "Groups" atau "TA/TO" (Travel Agency/ Tour Operator).
distribution_channel: Saluran distribusi yang digunakan untuk pemesanan, seperti "TA/TO" (Travel Agency/ Tour Operator) atau "Offline".
reserved_room_type: Jenis kamar yang dipesan oleh tamu.
assigned_room_type: Jenis kamar yang diberikan kepada tamu saat kedatangan.
deposit_type: Jenis deposit yang digunakan, misalnya "Non Refund" (tidak dapat dikembalikan).
customer_type: Jenis pelanggan, seperti "Transient" yang menunjukkan pelanggan sementara.
reservation_status: Status pemesanan, seperti "Canceled" (Dibatalkan) atau "Check-Out" (Sudah Check-out).
reservation_status_date: Tanggal status pemesanan diupdate atau tercatat.

```

### numerical variabel

In [ ]:
numerical.head()

is_canceled  lead_time  arrival_date_year  arrival_date_week_number  \
0            0        342               2015                        27   
1            0        737               2015                        27   
2            0          7               2015                        27   
3            0         13               2015                        27   
4            0         14               2015                        27   

   arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  \
0                          1                        0                     0   
1                          1                        0                     0   
2                          1                        0                     1   
3                          1                        0                     1   
4                          1                        0                     2   

   adults  children  babies  is_repeated_guest  previous_cancellations  \
0       2       0.0       0                  0                       0   
1       2       0.0       0                  0                       0   
2       1       0.0       0                  0                       0   
3       1       0.0       0                  0                       0   
4       2       0.0       0                  0                       0   

   previous_bookings_not_canceled  booking_changes  agent  company  \
0                               0                3    NaN      NaN   
1                               0                4    NaN      NaN   
2                               0                0    NaN      NaN   
3                               0                0  304.0      NaN   
4                               0                0  240.0      NaN   

   days_in_waiting_list   adr  required_car_parking_spaces  \
0                     0   0.0                            0   
1                     0   0.0                            0   
2                     0  75.0                            0   
3                     0  75.0                            0   
4                     0  98.0                            0   

   total_of_special_requests  
0                          0  
1                          0  
2                          0  
3                          0  
4                          1

Ada banyak nilai nol pada numerical variabel, di mana hal itu biasanya menunjukkan adanya missing value dalam data. Akan tetapi, jika melakukan pendekatan berdasarkan konteks yang ada, di mana variabel is_canceled sebagai salah satu variabel yang memiliki banyak nilai nol, tentunya tidak masuk akal jika nilai nol mengindikasikan jumlah data yang hilang. Lebih logis jika nilai nol tersebut memang ada karena merupakan nilai yang sah.

Jika melihat variabel di sebelahnya pada 'data', yakni hotel dan lead_time. Pendekatan konteks serupa juga dapat dilakukan:
* Hotel: tidak mungkin mengandung nilai nol karena nilai-nilainya berupa kategori (teks) yang menunjukkan tipe hotel. Setiap entri pada variabel ini seharusnya berisi informasi hotel yang valid tanpa nilai nol, sehingga nol tidak menjadi masalah di sini.
* Lead_time: berdasarkan konteksnya, Variabel ini menunjukkan waktu (dalam hari) antara saat pemesanan dilakukan dan tanggal kedatangan. Jika ada nilai nol pada lead_time, kemungkinan besar artinya tamu melakukan pemesanan pada hari yang sama dengan kedatangan, dengan kata lain tidak ada selisih (hari) antara tanggal pemesanan dan kedatangan. Jadi, nilai nol di sini valid dan bukan indikasi missing value.

* Kemudian, variabel is_canceled sendiri secara kontekstualnya dapat terkait dengan pembatalan pesanan. Penambahan awalan is pada nama variabel juga biasanya ditujukan untuk pertanyaan yang mengharapkan 2 jawaban langsung seperti 'ya' atau 'tidak'. Dan umumnya penggunaan nilai biner pada bidang data, maka dapat diasumsikan bahwan ilai nol pada variabel ini juga bukan missing value, tetapi informasi sah yang menyatakan bahwa pemesanan berjalan sesuai rencana tanpa dibatalkan.  Nilai 0 di sini secara logis menunjukkan status "Tidak" (tidak dibatalkan), sedangkan 1 menunjukkan "Ya" (dibatalkan).

In [ ]:
for col in numerical.columns:
    print("===============================")
    print(numerical[col].value_counts())
    print("\n")

is_canceled
0    75166
1    44224
Name: count, dtype: int64


lead_time
0      6345
1      3460
2      2069
3      1816
4      1715
       ... 
400       1
370       1
532       1
371       1
463       1
Name: count, Length: 479, dtype: int64


arrival_date_year
2016    56707
2017    40687
2015    21996
Name: count, dtype: int64


arrival_date_week_number
33    3580
30    3087
32    3045
34    3040
18    2926
21    2854
28    2853
17    2805
20    2785
29    2763
42    2756
31    2741
41    2699
15    2689
27    2664
25    2663
38    2661
23    2621
35    2593
39    2581
22    2546
24    2498
13    2416
16    2405
19    2402
40    2397
26    2391
43    2352
44    2272
14    2264
37    2229
8     2216
36    2167
10    2149
9     2117
7     2109
12    2083
11    2070
45    1941
53    1816
49    1782
47    1685
46    1574
6     1508
50    1505
48    1504
4     1487
5     1387
3     1319
2     1218
52    1195
1     1047
51     933
Name: count, dtype: int64


arrival_date_day_of_month
17   

```
is_canceled: Menunjukkan apakah pemesanan dibatalkan (1) atau tidak (0).
lead_time: Waktu antara saat pemesanan dilakukan dan tanggal kedatangan yang dijadwalkan (dalam hari).
arrival_date_year: Tahun kedatangan yang dijadwalkan.
arrival_date_week_number: Minggu ke-berapa dari tahun kedatangan yang dijadwalkan (1-52).
arrival_date_day_of_month: Tanggal kedatangan dalam bulan (1-31).
stays_in_weekend_nights: Jumlah malam akhir pekan yang akan diinap oleh tamu (Sabtu atau Minggu).
stays_in_week_nights: Jumlah malam hari kerja yang akan diinap oleh tamu (Senin-Jumat).
adults: Jumlah orang dewasa dalam pemesanan.
children: Jumlah anak-anak dalam pemesanan.
babies: Jumlah bayi dalam pemesanan.
is_repeated_guest: Menunjukkan apakah tamu adalah tamu berulang (1) atau tamu baru (0).
previous_cancellations: Jumlah pemesanan sebelumnya yang dibatalkan oleh tamu.
previous_bookings_not_canceled: Jumlah pemesanan sebelumnya yang tidak dibatalkan.
booking_changes: Jumlah perubahan yang dilakukan pada pemesanan setelah awalnya dikonfirmasi.
agent: Kode agen perjalanan yang melakukan pemesanan (mungkin berisi nilai NaN jika tidak ada agen yang terlibat).
company: Kode perusahaan yang bertanggung jawab atas pemesanan (mungkin berisi nilai NaN jika tidak ada perusahaan yang terlibat).
days_in_waiting_list: Jumlah hari pemesanan menunggu dalam daftar tunggu sebelum dikonfirmasi.
adr: Average Daily Rate atau harga harian rata-rata, dihitung dari total jumlah yang dibayarkan untuk pemesanan dibagi dengan jumlah malam menginap.
required_car_parking_spaces: Jumlah tempat parkir yang diminta oleh tamu.
total_of_special_requests: Jumlah permintaan khusus yang diminta oleh tamu (misalnya kamar di lantai tertentu, tempat tidur bayi, dll.).

```

## missing value checked

In [ ]:
# persebaran nilai nan dalam dataset
missing = data.isnull().sum().sort_values(ascending=False)
percent = (missing/len(data)* 100).sort_values(ascending=False)
missing_value = pd.concat([missing, percent], axis=1, keys=['total', 'percent'])
missing_value

total    percent
company                         112593  94.306893
agent                            16340  13.686238
country                            488   0.408744
children                             4   0.003350
reserved_room_type                   0   0.000000
assigned_room_type                   0   0.000000
booking_changes                      0   0.000000
deposit_type                         0   0.000000
hotel                                0   0.000000
previous_cancellations               0   0.000000
days_in_waiting_list                 0   0.000000
customer_type                        0   0.000000
adr                                  0   0.000000
required_car_parking_spaces          0   0.000000
total_of_special_requests            0   0.000000
reservation_status                   0   0.000000
previous_bookings_not_canceled       0   0.000000
is_repeated_guest                    0   0.000000
is_canceled                          0   0.000000
distribution_channel                 0   0.000000
market_segment                       0   0.000000
meal                                 0   0.000000
babies                               0   0.000000
adults                               0   0.000000
stays_in_week_nights                 0   0.000000
stays_in_weekend_nights              0   0.000000
arrival_date_day_of_month            0   0.000000
arrival_date_week_number             0   0.000000
arrival_date_month                   0   0.000000
arrival_date_year                    0   0.000000
lead_time                            0   0.000000
reservation_status_date              0   0.000000

Seperti hasil yang didapatkan dari general information, terdapat 4 variabel yang memiliki missing value. Dengan variabel company yang 90% lebih datanya mengandung nilai yang hilang atau kosong. Selanjutnya, bagaimana sampai ada missing value dalam dataset?

Apakah karena memang sudah ada sejak awal atau terdapat kesalahan input dari pengguna? Error sistem? atau karena hal lain?

In [ ]:
nan_variabel = data[data.isnull().any(axis=1)]
nan_variabel.sample(50)

hotel  is_canceled  lead_time  arrival_date_year  \
77125     City Hotel            0        188               2015   
113122    City Hotel            0        409               2017   
79831     City Hotel            0         38               2015   
112934    City Hotel            0        137               2017   
76847     City Hotel            0        204               2015   
19775   Resort Hotel            0         41               2015   
18673   Resort Hotel            0          0               2015   
115523    City Hotel            0        146               2017   
111394    City Hotel            0        210               2017   
72823     City Hotel            1        111               2017   
31810   Resort Hotel            0          1               2017   
76753     City Hotel            0          0               2015   
81365     City Hotel            0         11               2015   
80991     City Hotel            0         24               2015   
112044    City Hotel            0         33               2017   
93409     City Hotel            0        155               2016   
35642   Resort Hotel            0        252               2017   
37718   Resort Hotel            0        273               2017   
25826   Resort Hotel            0          2               2016   
18173   Resort Hotel            1        289               2015   
51730     City Hotel            1        158               2016   
108220    City Hotel            0          1               2017   
106951    City Hotel            0        203               2017   
24176   Resort Hotel            0        468               2016   
108027    City Hotel            0         49               2017   
49246     City Hotel            0         94               2016   
101723    City Hotel            0          8               2016   
113840    City Hotel            0         73               2017   
52965     City Hotel            1         71               2016   
5260    Resort Hotel            1        106               2016   
2730    Resort Hotel            0          1               2015   
39237   Resort Hotel            0        174               2017   
44624     City Hotel            0         80               2015   
30150   Resort Hotel            0         97               2016   
104702    City Hotel            0         29               2017   
98228     City Hotel            0          1               2016   
32773   Resort Hotel            0          4               2017   
91602     City Hotel            0         69               2016   
111633    City Hotel            0        120               2017   
16685   Resort Hotel            0          0               2015   
49115     City Hotel            1         55               2016   
117618    City Hotel            0        139               2017   
76301     City Hotel            0        264               2015   
81838     City Hotel            1        178               2016   
93697     City Hotel            0          4               2016   
84638     City Hotel            0        116               2016   
98634     City Hotel            0        256               2016   
76972     City Hotel            1        342               2015   
15      Resort Hotel            0         68               2015   
3153    Resort Hotel            1        122               2015   

       arrival_date_month  arrival_date_week_number  \
77125           September                        37   
113122                May                        22   
79831             October                        44   
112934                May                        22   
76847              August                        33   
19775            December                        53   
18673            November                        48   
115523               July                        27   
111394                May                        19   
72823              August                      

Pertama-tama, mari lihat persebaran data tidak nan-nya terlebih dahulu:

In [ ]:
not_nan_variable = data[data.notnull().all(axis=1)]
not_nan_variable.sample(50)

hotel  is_canceled  lead_time  arrival_date_year  \
18270   Resort Hotel            0         12               2015   
43580     City Hotel            0         19               2015   
93017     City Hotel            0        131               2016   
3058    Resort Hotel            0         33               2015   
104473    City Hotel            0          1               2017   
15263   Resort Hotel            0         24               2015   
2969    Resort Hotel            0         37               2015   
3061    Resort Hotel            0         21               2015   
37492   Resort Hotel            0        222               2017   
18855   Resort Hotel            0          0               2015   
3051    Resort Hotel            0         33               2015   
21696   Resort Hotel            0         17               2016   
3078    Resort Hotel            0         15               2015   
18201   Resort Hotel            0         39               2015   
3015    Resort Hotel            0         40               2015   
89689     City Hotel            0         11               2017   
18284   Resort Hotel            0         36               2015   
25635   Resort Hotel            0          9               2016   
18195   Resort Hotel            0         21               2015   
105447    City Hotel            0         20               2017   
18283   Resort Hotel            0         24               2015   
3043    Resort Hotel            0         41               2015   
18265   Resort Hotel            0         36               2015   
18202   Resort Hotel            0         36               2015   
18285   Resort Hotel            0         24               2015   
3037    Resort Hotel            0         33               2015   
2697    Resort Hotel            0         24               2015   
18149   Resort Hotel            0         23               2015   
18316   Resort Hotel            0         31               2015   
18221   Resort Hotel            0         24               2015   
89690     City Hotel            0         14               2017   
99952     City Hotel            0        256               2016   
19901   Resort Hotel            0          8               2016   
18217   Resort Hotel            0         37               2015   
21080   Resort Hotel            1          4               2016   
2881    Resort Hotel            0         24               2015   
3004    Resort Hotel            1         40               2015   
18280   Resort Hotel            0         24               2015   
14090   Resort Hotel            0          5               2015   
32641   Resort Hotel            0         18               2017   
113627    City Hotel            0        210               2017   
21079   Resort Hotel            0         10               2016   
99964     City Hotel            0        256               2016   
18297   Resort Hotel            0         31               2015   
8762    Resort Hotel            0        364               2016   
18227   Resort Hotel            0         31               2015   
2917    Resort Hotel            0         27               2015   
31759   Resort Hotel            0          5               2017   
18264   Resort Hotel            0         19               2015   
2918    Resort Hotel            0         35               2015   

       arrival_date_month  arrival_date_week_number  \
18270            November                        45   
43580           September                        39   
93017                July                        29   
3058             November                        46   
104473            January                         2   
15263             October                        44   
2969             November                        46   
3061             November                        46   
37492                June                        24   
18855            November                      

Dari kedua table di atas, variabel yang memiliki nilai nan dapat dibandingkan dengan yang tidak memiliki nilai nan, yang menghasilkan analisis berikut:

####1) **Why does the Company have missing value?**

* Dibandingkan dengan data yang memiliki nilai nan, data yang tidak memiliki nilai nan memiliki persentase yang lebih kecil. Hal ini jika dilihat sekilas tentu saja benar. Akan tetapi, banyaknya nilai nan itu sendirilah masalahnya.
* Pada variabel company, nilainya diisi dengan kode perusahaan yang melakukan pemesanan. Hal ini seperti "apakah (tamu) yang memesan adalah sebuah perusahaan?". Jika demikian, "jika pada variabel company diisi dengan yang bukan merupakan missing value, apakah berarti yang memesan reservasi adalah atas nama perusahaan?"
* Jika diamati, persebaran tersebut dapat diperkecil ruang lingkupnya dengan melihat hubungannya dengan variabel lain. Seperti, "siapa-siapa saja tamunya?". Hal tersebut dapat diketahui dengan melihat variabel children dan babies, dengan asumsi 'jika tamunya berasal dari perusahaan atau kalangan (orang) bisnis (contoh: pekerja kantoran)' maka secara logika tamu-tamu tersebut tidak akan membawa keluarga mereka karena harus fokus pada pekerjaannya, kecuali hal itu memungkinkan. Kemudian, pada variabel adults dan market_segment hal itu diperjelas dengan hanya sedikit orang dewasa yang merupakan tamu dari perusahaan di mana sebagian besar berasal dari segment 'coorporate'. Artinya, berpergian sendiri.

####2) **Why does the agent have missing value?**

In [ ]:
nan_variabel = data[data['agent'].isnull()]
nan_variabel.sample(50)

hotel  is_canceled  lead_time  arrival_date_year  \
74719     City Hotel            0          0               2016   
68635     City Hotel            1        105               2017   
14005   Resort Hotel            0          8               2015   
105982    City Hotel            0        114               2017   
24296   Resort Hotel            0          7               2016   
23104   Resort Hotel            0         58               2016   
20570   Resort Hotel            0          2               2016   
12860   Resort Hotel            1          2               2017   
96751     City Hotel            0          9               2016   
64211     City Hotel            1         28               2017   
71063     City Hotel            1        141               2017   
19732   Resort Hotel            0        112               2015   
14756   Resort Hotel            1        259               2015   
6150    Resort Hotel            0        225               2016   
115478    City Hotel            0         34               2017   
14691   Resort Hotel            1        259               2015   
47546     City Hotel            1         35               2016   
84138     City Hotel            0         99               2016   
32132   Resort Hotel            0          3               2017   
21495   Resort Hotel            0          0               2016   
64127     City Hotel            1         28               2017   
39138   Resort Hotel            0         51               2017   
3499    Resort Hotel            1         93               2015   
65544     City Hotel            1         54               2017   
69161     City Hotel            1        168               2017   
10242   Resort Hotel            1         70               2017   
64147     City Hotel            1         28               2017   
22699   Resort Hotel            0        169               2016   
1418    Resort Hotel            1         35               2015   
76716     City Hotel            1          8               2016   
21136   Resort Hotel            0          4               2016   
113814    City Hotel            0         27               2017   
4100    Resort Hotel            1        140               2016   
90464     City Hotel            0          0               2016   
21171   Resort Hotel            0          0               2016   
21134   Resort Hotel            0          0               2016   
373     Resort Hotel            0        109               2015   
31567   Resort Hotel            0         62               2016   
37453   Resort Hotel            0         36               2017   
22726   Resort Hotel            0        129               2016   
23268   Resort Hotel            0         32               2016   
15406   Resort Hotel            0        185               2015   
61019     City Hotel            1         42               2016   
78481     City Hotel            0         20               2015   
34260   Resort Hotel            0        112               2017   
64735     City Hotel            1         92               2017   
55557     City Hotel            1        138               2016   
17635   Resort Hotel            0        287               2015   
92059     City Hotel            0          0               2016   
76897     City Hotel            0          0               2015   

       arrival_date_month  arrival_date_week_number  \
74719                 May                        19   
68635                 May                        20   
14005            November                        48   
105982           February                         7   
24296                 May                        21   
23104               April                        16   
20570               March                        10   
12860                July                        29   
96751              August                        36   
64211               March                      

* Pertama, agent tidak setinggi company dalam persebaran missing value-nya, tetapi tetap menjadi perhatian utama. Seperti misalnya, ketika pada data yang tidak memiliki nan, sebagian besar data segment dan channel berpusat pada coorporate atau TA/TO, tetapi group, direct dan pilihan pemesanan sendiri tanpa melalui agent ketika berada di data yang memiliki missing value. Hal ini, tentu memberikan petunjuk bahwa sebagian besar tamu memiliki preferensi yang berbeda dalam melakukan reservasi kamar hotel.
* Di sisi lain, pemesanan di mana variabel agent-nya adalah missing value, dapat diartikan bahwa tamu/pelanggan tidak dalam jangkauan dari (kode) agent terkait,  yang mengindikasikan bahwa pelanggan memang lebih menyukai melakukan reservasi secara mandiri dan sistem agent hotel yang kurang dalam menjangkau pelanggan.

####3) **Why does the country have missing value?**

In [ ]:
nan_variabel = data[data['country'].isnull()]
nan_variabel.sample(50)

hotel  is_canceled  lead_time  arrival_date_year  \
13811   Resort Hotel            1         71               2016   
14040   Resort Hotel            0          0               2016   
14362   Resort Hotel            0          0               2015   
14053   Resort Hotel            0          5               2015   
13919   Resort Hotel            0          1               2015   
14005   Resort Hotel            0          8               2015   
14295   Resort Hotel            0         17               2016   
14556   Resort Hotel            0         66               2016   
14496   Resort Hotel            0         40               2017   
14076   Resort Hotel            0          0               2016   
14366   Resort Hotel            0          1               2016   
14167   Resort Hotel            0          6               2015   
14426   Resort Hotel            0          1               2016   
14516   Resort Hotel            0          4               2017   
14476   Resort Hotel            0          0               2015   
14655   Resort Hotel            0        266               2015   
13977   Resort Hotel            0          0               2015   
9612    Resort Hotel            1          4               2016   
14481   Resort Hotel            0          1               2016   
13816   Resort Hotel            0          6               2016   
14293   Resort Hotel            0          8               2016   
13907   Resort Hotel            0          0               2016   
14131   Resort Hotel            0          0               2016   
65910     City Hotel            1          0               2017   
14639   Resort Hotel            0          7               2016   
14629   Resort Hotel            0          3               2016   
14498   Resort Hotel            0          1               2015   
14604   Resort Hotel            0          1               2016   
14251   Resort Hotel            0         35               2016   
14009   Resort Hotel            0          0               2016   
13834   Resort Hotel            0          1               2016   
13883   Resort Hotel            0          0               2017   
14194   Resort Hotel            0          2               2016   
15341   Resort Hotel            1        246               2015   
101488    City Hotel            0          1               2016   
14157   Resort Hotel            0          2               2015   
14522   Resort Hotel            0          0               2016   
14360   Resort Hotel            0          0               2016   
14308   Resort Hotel            0          3               2016   
14564   Resort Hotel            0          2               2017   
14478   Resort Hotel            0          0               2016   
15178   Resort Hotel            1        246               2015   
14605   Resort Hotel            0          3               2017   
14104   Resort Hotel            0        167               2016   
13921   Resort Hotel            0          0               2016   
13948   Resort Hotel            0          0               2015   
14420   Resort Hotel            0          0               2016   
14575   Resort Hotel            0          1               2016   
14085   Resort Hotel            1          2               2016   
14671   Resort Hotel            0          3               2016   

       arrival_date_month  arrival_date_week_number  \
13811                July                        29   
14040            February                         6   
14362            December                        50   
14053            November                        48   
13919            November                        45   
14005            November                        48   
14295            December                        50   
14556            December                        53   
14496             January                         3   
14076            December                      

* Berdasarkan definisi country yang telah diambil sebelumnya, pada variabel ini adanya country dapat menunjukkan bahwa 'tamu tersebut dari negara b' atau 'negara asal seorang tamu melakukan reservasi', tetapi tidak adanya nilai pada variabel country tidak dapat diartikan bahwa 'tamu tersebut tidak berasal dari negara manapun' atau 'tamu tersebut tidak memiliki negara asal'. Bahkan jika terdapat [negara-negara yang tidak diakui oleh PBB](https://worldpopulationreview-com.translate.goog/country-rankings/countries-not-in-the-un?_x_tr_sl=en&_x_tr_tl=id&_x_tr_hl=id&_x_tr_pto=tc&_x_tr_hist=true), hal tersebut tidak dapat dijadikan patokan sebab konsep negara asal tetap berlaku dan tidak dapat dijadikan acuan pasti dalam analisis variabel ini.
* Karena itu, kemungkinan adanya tamu gelap dapat dihilangkan dan adanya tamu yang baik secara tidak atau dsengaja tidak mengisi negara asalnya juga dapat dipertimbangkan. Selain itu, jika tamu memesan reservasi melalui agent atau layanan bantuan pihak ketiga yang sejenisnya dan variabel country tetap tidak terisi, maka terdapat error dalam sistem pemesanan dan penginputan— karena terkait preferensi pelanggan/tamu yang mungkin saja 'ingin' menjadi annonim.

####4) **Why does the children have missing value?**

In [ ]:
nan_variabel = data[data['children'].isnull()]
nan_variabel

hotel  is_canceled  lead_time  arrival_date_year  \
40600  City Hotel            1          2               2015   
40667  City Hotel            1          1               2015   
40679  City Hotel            1          1               2015   
41160  City Hotel            1          8               2015   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
40600             August                        32                          3   
40667             August                        32                          5   
40679             August                        32                          5   
41160             August                        33                         13   

       stays_in_weekend_nights  stays_in_week_nights  adults  children  \
40600                        1                     0       2       NaN   
40667                        0                     2       2       NaN   
40679                        0                     2       3       NaN   
41160                        2                     5       2       NaN   

       babies meal country market_segment distribution_channel  \
40600       0   BB     PRT      Undefined            Undefined   
40667       0   BB     PRT         Direct            Undefined   
40679       0   BB     PRT      Undefined            Undefined   
41160       0   BB     PRT      Online TA            Undefined   

       is_repeated_guest  previous_cancellations  \
40600                  0                       0   
40667                  0                       0   
40679                  0                       0   
41160                  0                       0   

       previous_bookings_not_canceled reserved_room_type assigned_room_type  \
40600                               0                  B                  B   
40667                               0                  B                  B   
40679                               0                  B                  B   
41160                               0                  B                  B   

       booking_changes deposit_type  agent  company  days_in_waiting_list  \
40600                0   No Deposit    NaN      NaN                     0   
40667                0   No Deposit   14.0      NaN                     0   
40679                0   No Deposit    NaN      NaN                     0   
41160                0   No Deposit    9.0      NaN                     0   

         customer_type   adr  required_car_parking_spaces  \
40600  Transient-Party  12.0                            0   
40667  Transient-Party  12.0                            0   
40679  Transient-Party  18.0                            0   
41160  Transient-Party  76.5                            0   

       total_of_special_requests reservation_status reservation_status_date  
40600                          1           Canceled                8/1/2015  
40667                          1           Canceled                8/4/2015  
40679                          2           Canceled                8/4/2015  
41160                          1           Canceled                8/9/2015

* Pada variabel children yang memiliki nilai nan, keempatnya merupakan pemesanan yang dibatalkan di selang waktu yang berdekatan. Selain itu, status reservasi nya pun di batalkan. Hal tersebut menunjukkan bahwa mungkin saja terjadi human error saat pelanggan melakukan reservasi. Kemungkinan itu tidak jauh dari definisi children 'adakah pelanggan yang merupakan bayi dalam daftar reservasi?'.
* Selain itu, nilai NaN pada variabel ini dapat disebabkan oleh pelanggan yang tidak mengisi informasi children karena reservasi dibatalkan lebih awal sebelum data diinput secara lengkap. Faktor lainnya adalah kurangnya validasi data dalam sistem selama proses pengisian formulir reservasi.

In [ ]:
data['children'].describe()

count    119386.000000
mean          0.103890
std           0.398561
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max          10.000000
Name: children, dtype: float64

## duplicated value checked

In [ ]:
data.duplicated().sum()

31994

In [ ]:
duplicates_variabel = data.apply(lambda x: x.duplicated().sum())
duplicates_variabel

hotel                             119388
is_canceled                       119388
lead_time                         118911
arrival_date_year                 119387
arrival_date_month                119378
arrival_date_week_number          119337
arrival_date_day_of_month         119359
stays_in_weekend_nights           119373
stays_in_week_nights              119355
adults                            119376
children                          119384
babies                            119385
meal                              119385
country                           119212
market_segment                    119382
distribution_channel              119385
is_repeated_guest                 119388
previous_cancellations            119375
previous_bookings_not_canceled    119317
reserved_room_type                119380
assigned_room_type                119378
booking_changes                   119369
deposit_type                      119387
agent                             119056
company                           119037
days_in_waiting_list              119262
customer_type                     119386
adr                               110511
required_car_parking_spaces       119385
total_of_special_requests         119384
reservation_status                119387
reservation_status_date           118464
dtype: int64


___
## **Conclusion so far ~**


| No. | Variabel | Problem | Deskripsi Masalah | Tindakan |
|-----|-----|------|--------|------|
| 1. | reservation_status_date | Kesalahan tipedata: object | b | Ubah ke dalam tipedata yang benar|
| 2. | company | Missing value: Non-Applicable Data | Variabel tidak berlaku untuk semua kasus,<br>hanya relevan jika pemesannya perusahaan| imputasi dengan nilai 0; tidak dari<br>perusahaan manapun|
| 3. | agent | Missing value: Non-Applicable Data | Variabel tidak berlaku untuk semua kasus,<br> tidak semua pesanan membutuhkan agent| Imputasi dengan nilai 0; tidak dari<br>agent manapun|
| 4. | country | Missing value: Kesalahan pencatatan<br>dan teknis/sistem | User tidak mengisi pemesanan reservasi<br>dengan benar, kurangnya verifikasi sistem<br>terkait informasi penting sebelum<br>melakukan reservasi | Imputasi dengan mean
| 5. | children | Human error dan kurangnya verifikasi sistem | melakukan pembatalan sebelum mengisi<br>'jumlah tamu yang merupakan anak-anak'|
| 6. |  | | |


___

# 3) cleaning data

## handling missing value

In [ ]:
data['company'] = data['company'].fillna(0)
data['agent'] = data['agent'].fillna(0)
data['country'] = data['country'].fillna(0)
data['children'] = data['children'].fillna(0)

In [ ]:
missing_value

total  percent
hotel                               0      0.0
is_canceled                         0      0.0
reservation_status                  0      0.0
total_of_special_requests           0      0.0
required_car_parking_spaces         0      0.0
adr                                 0      0.0
customer_type                       0      0.0
days_in_waiting_list                0      0.0
company                             0      0.0
agent                               0      0.0
deposit_type                        0      0.0
booking_changes                     0      0.0
assigned_room_type                  0      0.0
reserved_room_type                  0      0.0
previous_bookings_not_canceled      0      0.0
previous_cancellations              0      0.0
is_repeated_guest                   0      0.0
distribution_channel                0      0.0
market_segment                      0      0.0
country                             0      0.0
meal                                0      0.0
babies                              0      0.0
children                            0      0.0
adults                              0      0.0
stays_in_week_nights                0      0.0
stays_in_weekend_nights             0      0.0
arrival_date_day_of_month           0      0.0
arrival_date_week_number            0      0.0
arrival_date_month                  0      0.0
arrival_date_year                   0      0.0
lead_time                           0      0.0
reservation_status_date             0      0.0

## handling duplicated value

In [ ]:
data.drop_duplicates(inplace=True)
data.duplicated().sum()

0

### Inaccurate data

In [ ]:
pd.to_datetime(data['reservation_status_date'])
print("tipedata sudah diperbaiki")

tipedata sudah diperbaiki


# Hasil yang di dapat

In [ ]:
!git --version

git version 2.34.1


In [ ]:
!git clone https://github.com/Lanlanb/Studycase.git

Cloning into 'Studycase'...
remote: Enumerating objects: 12, done.
remote: Counting objects: 100% (12/12), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 12 (delta 2), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (12/12), 1.21 MiB | 5.42 MiB/s, done.
Resolving deltas: 100% (2/2), done.


In [ ]:
!ls /content

sample_data  Studycase


*harap diingat bahwa beberapa bagian tidak ditampilkan karena berkenan dengan privasi dan akses pengguna

In [ ]:
datasets = {
    'data' : data,
    'categorical' : categorical,
    'numerical' : numerical
}

In [ ]:
for name, data in datasets.items():
    path = f"/content/Studycase/Hotel_bookings/{name}.csv"
    data.to_csv(path, index=False)
    print(f"{name} berhasil di simpan di {path}.")

data berhasil di simpan di /content/Studycase/Hotel_bookings/data.csv.
categorical berhasil di simpan di /content/Studycase/Hotel_bookings/categorical.csv.
numerical berhasil di simpan di /content/Studycase/Hotel_bookings/numerical.csv.


**Kontak Saya**:

___
Email: contact@nurlaelabulan.my.id

LinkedIn: [bulan-nurlaela](https://linkedin.com/in/bulan-nurlaela)

Github: [lanlanb](https://github.com/Lanlanb)